src/Security/MemberAuthenticator/LoginHandler.php

```diff
@@ -263,7 +264,14 @@ protected function redirectToChangePassword()
             'good'
         );
         $changedPasswordLink = Security::singleton()->Link('changepassword');
+        $changePasswordUrl = $this->addBackURLParam($changedPasswordLink);
+        $whitelistedUrl = Controller::join_links($changedPasswordLink, 'ChangePasswordForm');
 
-        return $this->redirect($this->addBackURLParam($changedPasswordLink));
+        $session = $this->getRequest()->getSession();
+        $passwordExpirationMiddleware = Injector::inst()->get(PasswordExpirationMiddleware::class);
+        $passwordExpirationMiddleware->setRedirect($session, $changePasswordUrl);
+        $passwordExpirationMiddleware->whitelistUrl($session, $whitelistedUrl);
```

Comment:
I'm still not convinced we really need both the explicit URL and the partially matched URL - if you just use partial matching, you can probably remove all of this coupling, right? Then you just use the config options for what the changepassword URL is, and what the allowed URLs can start with

Reply:
yeah, I guess that might be superfluous in our circumstances. Will remove in sake of reducing coupling and introduced API surface.

Reply:
done